src/ATen/native/sparse/xpu/sycl/SparseTensorMathKernel.cpp

```diff
@@ -0,0 +1,388 @@
+
+  auto num_groups = CeilDiv(nnz, group_size);
+  if (num_groups > max_work_group_num)
+    num_groups = max_work_group_num;
+  auto total_items = num_groups * group_size;
+
+  KernelClass kfn(op, dense, indices, values, nnz);
+  sycl_kernel_submit(total_items, group_size, queue, kfn);
+}
+
+template <typename Op, typename IndexType, typename Real>
+struct SparseElementwiseKernelScalarFunctor {
+  void operator()(sycl::nd_item<1> item) const {
+    IndexType ind_skip = indices.strides[0];
+    IndexType ind_nnz_skip = indices.strides[1];
+    IndexType value_skip = values.strides[0];
+
+    for (IndexType linearId = (IndexType)item.get_group_linear_id() *
+                 (IndexType)item.get_local_range()[0] +
+             (IndexType)item.get_local_id()[0];
+         linearId < nnz;
+         linearId += (IndexType)item.get_group_range()[0] *
+             (IndexType)item.get_local_range()[0]) {
+      IndexType index = 0;
+      for (IndexType d = 0; d < indices.sizes[0]; d++) {
+        index = dense.sizes[d] * index +
+            indices.data[d * ind_skip + linearId * ind_nnz_skip];
+      }
+      op(dense.data + index, values.data + linearId * value_skip);
+    }
+  }
+  SparseElementwiseKernelScalarFunctor(
+      Op op,
+      TensorInfo<Real, IndexType> dense_info,
+      TensorInfo<indexT, IndexType> indices_info,
+      TensorInfo<Real, IndexType> values_info,
+      const IndexType nnz_value)
+      : op(op),
+        dense(dense_info),
+        indices(indices_info),
+        values(values_info),
+        nnz(nnz_value) {}
+
+ private:
+  Op op;
+  TensorInfo<Real, IndexType> dense;
+  TensorInfo<indexT, IndexType> indices;
+  TensorInfo<Real, IndexType> values;
+  const IndexType nnz;
+};
+
+template <typename Op, typename IndexType, typename Real>
+void sparseElementwiseKernelScalar(
+    Op op,
+    TensorInfo<Real, IndexType> dense,
+    TensorInfo<indexT, IndexType> indices,
+    TensorInfo<Real, IndexType> values,
+    const IndexType nnz) {
+  using KernelClass = SparseElementwiseKernelScalarFunctor<Op, IndexType, Real>;
+  auto& queue = getCurrentSYCLQueue();
+  IndexType group_size = (IndexType)syclMaxWorkGroupSize<KernelClass>();
+  IndexType target_global_size = (IndexType)syclMaxWorkItemsPerTile();
+  auto max_work_group_num = target_global_size / group_size;
+
+  auto num_groups = CeilDiv(nnz * group_size, group_size);
+  if (num_groups > max_work_group_num)
+    num_groups = max_work_group_num;
+  auto total_items = num_groups * group_size;
+
+  KernelClass kfn(op, dense, indices, values, nnz);
+  sycl_kernel_submit(total_items, group_size, queue, kfn);
+}
+} // namespace apply
+
+template <typename T>
+struct TensorCAddOp {
+  TensorCAddOp(T v) : val(v) {}
+
+  void operator()(T* out, T* in) const {
+    *out += val * *in;
+  }
+
+  void operator()(T* out, T* in1, T* in2) const {
+    *out = *in1 + val * *in2;
+  }
+
+  T val;
+};
+
+TORCH_XPU_API void add_out_dense_sparse_kernel(
```

Comment:
Move the decorator to the related header.

Reply:
done